### PR TITLE
Fix inventory cache interface

### PR DESCRIPTION
**Pull request by s-hertel**
##### SUMMARY
There are a number of problems with the current cache interface, the primary ones being that the developer interface is terrible and it only handles file-backed cache plugins.

Some of the interface problems are that the cache was difficult to set up, get() was expected to raise a KeyError, and the cache had to be set a particular way to be reflected in the backing plugin.

This allows self._cache to be used as a dictionary - e.g. if you want a KeyError, use ``__getattr__`` instead of ``get``. It also updates the backing cache plugin with the value of self._cache if it has been modified after parse is called so the developer doesn't need to worry about the discrepancy between the cache object being exposed (a dictionary) and the permanent cache store (CacheModule).

Added tests and developer documentation.

##### ISSUE TYPE
- Bugfix Pull Request
- Feature Pull Request

### CODE REVIEW

File: lib/ansible/vars/manager.py
Comment (s-hertel) [inline lib/ansible/vars/manager.py]: DON'T MERGE - just want to run tests on shippable before I remove this. I'm also interested in opinions about deprecating FactCache. :-)

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/101019/60/tests

File: lib/ansible/vars/fact_cache.py
Comment (s-hertel) [inline lib/ansible/vars/fact_cache.py]: I'm not sure how this was intended to be used and therefore what the right fix is. The alternative solution for this (and [this](https://github.com/ansible/ansible/pull/50446/files#diff-a8385538c61f1318541f0276f3e1ec7bR278)) is https://github.com/ansible/ansible/pull/40105/files#diff-cb51550dcfdbd27a3f224be2de15e2e8R396. Either the config definition loading must happen first to work with config or class_only needs to be used so the definitions are loaded first.

File: test/units/vars/test_variable_manager.py
Comment (s-hertel) [inline test/units/vars/test_variable_manager.py:24]: I could remove the changes to this file, it's irrelevant to the feature I've added. I just noticed it was masking the real fact_cache and so the tests weren't failing when I broke it.

File: lib/ansible/vars/fact_cache.py
Comment (s-hertel) [inline lib/ansible/vars/fact_cache.py]: I'm not sure how class_only was intended to be used and therefore what the right fix is. The alternative solution for this (and [this](https://github.com/ansible/ansible/pull/50446/files#diff-a8385538c61f1318541f0276f3e1ec7bR278)) is https://github.com/ansible/ansible/pull/40105/files#diff-cb51550dcfdbd27a3f224be2de15e2e8R396. Either the config definition loading must happen first to work with config or class_only needs to be used so the definitions are loaded first.

Comment (bcoca) [reply lib/ansible/vars/fact_cache.py]: the use looks fine, I'm not sure why it is needed, did the previous code not raise the error correctly?

File: test/units/vars/test_variable_manager.py
Comment (bcoca) [reply test/units/vars/test_variable_manager.py:24]: we kind of mask it in variablemanager itself, but we should not in these tests

File: lib/ansible/vars/fact_cache.py
Comment (s-hertel) [reply lib/ansible/vars/fact_cache.py]: It looks like you fixed this in 960388143f086bd229b3d5f779b7b9e2b78c8bd7 by moving the config loading first so I can remove the class_only stuff now. `get_option` and `set_options` rely on the definitions being loaded before creating an instance of the plugin. Those were not necessarily loaded first (causing this error https://github.com/ansible/ansible/blob/465df0ef8d27d7fd534f521335db11490ebd43bc/lib/ansible/config/manager.py#L462) and that's why this was necessary.

**Comment by s-hertel** [comment]
@acozine Can I get a review on this from a documentation perspective, please?

**Comment by s-hertel** [comment]
Rebased for porting guide and doc fragment conflicts.

File: lib/ansible/plugins/cache/__init__.py
Comment (s-hertel) [inline lib/ansible/plugins/cache/__init__.py:105]: changed in become plugins

File: docs/docsite/rst/dev_guide/developing_inventory.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_inventory.rst]: Is this the only way to get caching? If so, I'd recommend stating this as a direction rather than "if/then" (because "if/then" suggests there's an alternative - "if not/then"):

"To enable caching, use the ``Cacheable`` base class and extend the inventory plugin documentation with the ``inventory_cache`` documentation fragment."

File: docs/docsite/rst/dev_guide/developing_inventory.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_inventory.rst]:
```suggestion
Next, load the cache plugin specified by the user to read from and update the cache. If your inventory plugin uses YAML based configuration files and the ``_read_config_data`` method, the cache plugin is loaded within that method. If your inventory plugin does not use ``_read_config_data``, you must load the cache explicitly with ``load_cache_plugin``.
```

File: docs/docsite/rst/dev_guide/developing_inventory.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_inventory.rst]: Information on why this needs to be done might help users here - for security? to enable two-way communication? Something like, "Before using the cache, you must retrieve a unique cache key . . . . method. This required step establishes a secure connection." (or whatever the actual reason).

File: docs/docsite/rst/dev_guide/developing_inventory.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_inventory.rst]: This paragraph contains good info, and a better introductory sentence would make it great. Unfortunately I'm not sure what the main topic is, so I'm not sure what to suggest. It might be: "Now that you've enabled caching and loaded the correct plugin, you can set up the flow of data between the cache and your inventory using the ``cache`` parameter of the ``parse`` method." Or it might be:  "Be careful with the ``cache`` parameter, which controls data flow into and out of the cache; you want to avoid overwriting inventory at the wrong times." Again, detailing why this information is important helps users (and me!) understand it.

File: docs/docsite/rst/dev_guide/developing_inventory.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_inventory.rst]: Here I'd either fill in the "why" or make this a list, like this:
```
You have three other cache methods available:
- ``set_cache_plugin`` forces the cache plugin to be set with the contents of ``self._cache``
- ``update_cache_if_changed`` sets the cache plugin only if ``self._cache`` has been modified
- ``clear_cache`` deletes the keys in ``self._cache`` from your cache plugin
```

**Review by acozine — COMMENTED** [review]
Lots of nice work here - thanks for adding all this documentation. This review makes some suggestions on the developing_inventory parts. I'll look at the plugin dev parts tomorrow.

File: docs/docsite/rst/dev_guide/developing_plugins.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_plugins.rst]:
```suggestion
Import cache plugins using the cache_loader so you can use ``self.set_options()`` and ``self.get_option(<option_name>)``. If you import a cache plugin directly in the code base, you can only access options via ``ansible.constants``, and you break the cache plugin's ability to be used by an inventory plugin.
```

File: docs/docsite/rst/dev_guide/developing_plugins.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_plugins.rst]:
```suggestion
If you use the ``BaseCacheModule``, you must implement the methods ``get``, ``contains``, ``keys``, ``set``, ``delete``, ``flush``, and ``copy``. The ``contains`` method should return a boolean that indicates if the key exists and has not expired. Unlike file-based caches, the ``get`` method does not raise a KeyError if the cache has expired.
```

File: docs/docsite/rst/dev_guide/developing_plugins.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_plugins.rst]:
```suggestion
If you use the ``BaseFileCacheModule``, you must implement ``_load`` and ``_dump`` methods that will be called from the base class methods ``get`` and ``set``.
```

File: docs/docsite/rst/dev_guide/developing_plugins.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_plugins.rst]:
```suggestion
If your cache plugin stores JSON, use ``AnsibleJSONEncoder`` in the ``_dump`` or ``set`` method  and ``AnsibleJSONDecoder`` in the ``_load`` or ``get`` method.
```

File: docs/docsite/rst/plugins/inventory.rst
Comment (acozine) [inline docs/docsite/rst/plugins/inventory.rst]:
```suggestion
If an inventory plugin supports caching, you can enable and set caching options for an individual YAML configuration source or for multiple inventory sources using environment variables or Ansible configuration files. If you enable caching for an inventory plugin without providing inventory-specific caching options, the inventory plugin will use fact-caching options. Here is an example of enabling caching for an individual YAML configuration file:
```

File: docs/docsite/rst/porting_guides/porting_guide_2.8.rst
Comment (acozine) [inline docs/docsite/rst/porting_guides/porting_guide_2.8.rst]:
```suggestion
* Supporting file-backed caching via self.cache is deprecated and will
```

File: docs/docsite/rst/porting_guides/porting_guide_2.8.rst
Comment (acozine) [inline docs/docsite/rst/porting_guides/porting_guide_2.8.rst]:
```suggestion
  be removed in Ansible 2.12. If you maintain an inventory plugin, update it to use ``self._cache`` as a dictionary. For implementation details, see
```

File: docs/docsite/rst/porting_guides/porting_guide_2.8.rst
Comment (acozine) [inline docs/docsite/rst/porting_guides/porting_guide_2.8.rst]:
```suggestion
  the :ref:`developer guide on inventory plugins<inventory_plugin_caching>`.
```

File: docs/docsite/rst/porting_guides/porting_guide_2.8.rst
Comment (acozine) [inline docs/docsite/rst/porting_guides/porting_guide_2.8.rst]:
```suggestion
* Importing cache plugins directly is deprecated and will be removed in Ansible 2.12. Use the plugin_loader
```

File: docs/docsite/rst/porting_guides/porting_guide_2.8.rst
Comment (acozine) [inline docs/docsite/rst/porting_guides/porting_guide_2.8.rst]:
```suggestion
  so direct options, environment variables, and other means of configuration can be reconciled using the config
```

File: docs/docsite/rst/porting_guides/porting_guide_2.8.rst
Comment (acozine) [inline docs/docsite/rst/porting_guides/porting_guide_2.8.rst]: This section seems awfully long for the porting guide. I'd move anything that isn't already covered in the `developing_inventory` page to there, and link to it from here. That way folks will be able to find this info easily even when 2.8 starts to fade in the rear-view mirror.

**Review by acozine — COMMENTED** [review]
Thanks @s-hertel for expanding the documentation on writing and maintaining inventory and cache plugins! I've added some more suggestions - let me know if they spark questions or raise concerns.

File: docs/docsite/rst/porting_guides/porting_guide_2.8.rst
Comment (s-hertel) [reply docs/docsite/rst/porting_guides/porting_guide_2.8.rst]: Good point!

File: docs/docsite/rst/dev_guide/developing_inventory.rst
Comment (s-hertel) [reply docs/docsite/rst/dev_guide/developing_inventory.rst]: I think I made this better. The topic is the flow of data/logic.

**Comment by s-hertel** [comment]
Rebased for porting guide conflict.

**Comment by s-hertel** [comment]
@FutureTecSystems It looks like you have some valuable pieces in lib/ansible/plugins/inventory/foreman.py in that PR. I think it might be out of the scope of this pull request. I've tried to limit my changes to only those necessary for getting cache plugins to be functional with inventory plugins, so optimizing foreman may be better as a follow up PR.

**Review by bcoca — APPROVED** [review]

File: docs/docsite/rst/dev_guide/developing_inventory.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_inventory.rst]:
```suggestion
Now that you've enabled caching, loaded the correct plugin, and retrieved a unique cache key, you can set up the flow of data between the cache and your inventory using the ``cache`` parameter of the ``parse`` method. This value comes from the inventory manager and indicates whether the inventory is being refreshed (such as via ``--flush-cache`` or the meta task ``refresh_inventory``). Although the cache shouldn't be used to populate the inventory when being refreshed, the cache should be updated with the new inventory if the user has enabled caching. You can use ``self._cache`` like a dictionary. The following pattern allows refreshing the inventory to work in conjunction with caching.
```

File: docs/docsite/rst/dev_guide/developing_plugins.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_plugins.rst]:
```suggestion
```

**Review by acozine — COMMENTED** [review]
Thanks @s-hertel for creating and updating the docs for this. Last docs changes suggested.

**Review by acozine — APPROVED** [review]

**Comment by AlanCoding** [comment]
I'm hitting this message

```
[DEPRECATION WARNING]: InventoryModule should utilize self._cache as a dict 
instead of self.cache. To set the self._cache dictionary, use self._cache[key] 
= value instead of self.cache.set(key, value). To force update the underlying 
cache plugin with the contents of self._cache before parse() is complete, call 
self.set_cache_plugin and it will use the self._cache dictionary to update the 
cache plugin. This feature will be removed in version 2.12. Deprecation 
warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
```

with the openstack inventory plugin.

What does this mean? Is this automatically worthy of creating a new issue in your issue queue, since it looks like the message is directed at the maintainers of the inventory plugin itself, and the openstack inventory plugin is a part of Ansible.

**Pull request merged**
